### ISSUE TITLE
Homomorphisms

### ISSUE THREAD

**Issue by james-d-mitchell** [semigroups/Semigroups#799]
This issue relates to a discussion about adding code to Semigroups for computing homomorphisms etc between semigroups. 

Things to do:
1. make the return value of `SemigroupHomomorphism/IsomorphismByImages` in PR #797 be an object in a new representation `IsSemigroupHomomorphism/IsomorphismByImages` (i.e. call `SemigroupHomomorphismByImages2NC` at the end)
2. Write the `SemigroupHomomorphismByImages2NC` function to create a new object in the representation `IsSemigroupHomomorphismByImages`.
2. Might want `IsSemigroupHomomorphism/IsomorphismByFunction` representation also.
3. Implement all the related functionality `ImageElm`, `PreImages` etc
4. Replace all instances of `MagmaIsomorphismByFunctionNC` and `MappingByFunction` in the package by your new functions.
1. Implement some new functionality specific to semigroup homomorphisms:
   a. `KernelOfSemigroupHomomorphism` which should return a congruence on the `Source` of the homomorphism
   b.  The associated quotient semigroup i.e. `Source(map) / KernelOfSemigroupHomomorphism(map);`

**Comment by chinmayanagpal** [comment]
for random homomorphisms, get congruence classes of a semigroup, and then get the quotient, get a transformation semigroup isomorphic to it, and you have a surjective homomorphism

**Comment by artemiskonst** [comment]
Turns out the issue for quotient semigroups was not entirely solved. For some reason ImageSource(hom) is a weird type when hom is a semigroup homomorphism from a semigroup to a quotient semigroup. I can't even make an isomorphic fp semigroup. Even though ImageSource(hom) is equal to T below, I can't use Factorization on it or IsomorphismFpSemigroup. I'm not sure how to write a general issue for this as it seems quite specific to this function. 

~~~
gap> S := Semigroup([ Transformation( [ 2, 1, 5, 1, 5 ] ), Transformation( [ 1, 1, 1, 5, 3 ] ), Transformation( [ 2, 5, 3, 5, 3 ] ) ]);;
gap> congs := CongruencesOfSemigroup(S);;
gap> cong := congs[3];;
gap> T := S / cong;;
gap> gens := GeneratorsOfSemigroup(S);;
gap> images := List(gens, gen -> CongruenceClassOfElement(cong, gen));;
gap> hom := SemigroupHomomorphismByImagesNC2(S, T, gens, images);;
gap> ImagesSource(hom);                      
<semigroup with 3 generators>
gap> Display(last);
Semigroup( [ {Transformation( [ 2, 1, 5, 1, 5 ] )}, {Transformation( [ 1, 1, 1, 5, 3 ] )}, {Transformation( [ 2, 5, 3, 5, 3 ] )} ] )
gap> ImagesSource(hom) = T; 
true
gap> Factorization(T, images[2]);
[ 2 ]
gap> Factorization(ImagesSource(hom), images[2]);
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `Factorization' on 2 arguments at /Users/artemiskonst/gap/lib/methsel2.g:249 called from
<function "HANDLE_METHOD_NOT_FOUND">( <arguments> )
 called from read-eval loop at *stdin*:18
type 'quit;' to quit to outer loop

gap> map := IsomorphismFpSemigroup(ImagesSource(hom));
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 2nd choice method found for `IsomorphismFpSemigroup' on 1 arguments at /Users/artemiskonst/gap/lib/methsel2.g:249 called from
<function "HANDLE_METHOD_NOT_FOUND">( <arguments> )
 called from read-eval loop at *stdin*:18
type 'quit;' to quit to outer loop
~~~

**Comment by chinmayanagpal** [comment]
should we wrap it in something like `SemigroupHomomorphismByFunction` to make it nicer to replace the `MagmaHomomorphismByFunctionNC` and `MappingByFunction`?

**Comment by chinmayanagpal** [comment]
currently in `SemigroupHomomorphismByImages`, the generators of the source need to be given in the same order that they were given in when generating it using `Semigroup`. Otherwise the checking of relations fails.

**Comment by chinmayanagpal** [comment]
Unsure about `UnderlyingGeneralMapping` and `UnderlyingRelation`

**Comment by james-d-mitchell** [comment]
This issue is resolved in #797

**Comment by james-d-mitchell** [comment]
Resolved in v5.0.0
